subcommand/server-acl-init/command_ent_test.go

```diff
@@ -0,0 +1,436 @@
+
+	// Check the ACL auth method is created in the expected namespace.
+	authMethodName := "release-name-consul-k8s-auth-method"
+	methods, _, err := consul.ACL().AuthMethodList(&api.QueryOptions{
+		Namespace: "dest",
+	})
+	require.NoError(err)
+	found := false
+	for _, m := range methods {
+		if m.Name == authMethodName {
+			found = true
+		}
+	}
+	require.True(found, fmt.Sprintf("no auth method with name %s found", authMethodName))
+
+	// Check the binding rule is as expected.
+	rules, _, err := consul.ACL().BindingRuleList(authMethodName, &api.QueryOptions{
+		Namespace: "dest",
+	})
+	require.NoError(err)
+	require.Len(rules, 1)
+}
+
+// Test the auth method and acl binding rule created when namespaces are enabled
+// and we're mirroring namespaces.
+func TestRun_ConnectInject_NamespaceMirroring(t *testing.T) {
+	t.Parallel()
+	k8s, testAgent := completeEnterpriseSetup(t, resourcePrefix)
+	defer testAgent.Stop()
+	setupK8sServiceAccount(t, k8s)
+	require := require.New(t)
+
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI:        ui,
+		clientset: k8s,
+	}
+	cmd.init()
+	args := []string{
+		"-server-label-selector=component=server,app=consul,release=" + releaseName,
+		"-resource-prefix=" + resourcePrefix,
+		"-k8s-namespace=" + ns,
+		"-expected-replicas=1",
+		"-create-inject-auth-method",
+		"-enable-namespaces",
+		"-enable-inject-k8s-namespace-mirroring",
+		"-inject-k8s-namespace-mirroring-prefix=prefix-",
+	}
+	responseCode := cmd.Run(args)
+	require.Equal(0, responseCode, ui.ErrorWriter.String())
+
+	bootToken := getBootToken(t, k8s, resourcePrefix)
+	consul, err := api.NewClient(&api.Config{
+		Address: testAgent.HTTPAddr,
+		Token:   bootToken,
+	})
+	require.NoError(err)
+
+	// Check the ACL auth method is as expected.
+	authMethodName := "release-name-consul-k8s-auth-method"
+	method, _, err := consul.ACL().AuthMethodRead(authMethodName, nil)
+	require.NoError(err)
+	require.NotNil(method, authMethodName+" not found")
+	require.Contains(method.Config, "MapNamespaces")
+	require.Contains(method.Config, "ConsulNamespacePrefix")
+	require.Equal(true, method.Config["MapNamespaces"])
+	require.Equal("prefix-", method.Config["ConsulNamespacePrefix"])
+
+	// Check the binding rule is as expected.
+	rules, _, err := consul.ACL().BindingRuleList(authMethodName, nil)
+	require.NoError(err)
+	require.Len(rules, 1)
+}
+
+// Test that re-running the commands results in auth method and binding rules
+// being updated.
+func TestRun_ConnectInject_Updates(t *testing.T) {
+	t.Parallel()
+
+	cases := map[string]struct {
+		// Args for first run of command.
+		FirstRunArgs []string
+		// Args for second run of command.
+		SecondRunArgs []string
+		// Expected namespace for the auth method.
+		AuthMethodExpectedNS string
+		// Whether to expect the MapNamespaces and ConsulNamespacePrefix
+		// keys to be set on the config.
+		AuthMethodExpectMirroringConfig bool
+		// Expected namespace for the binding rule.
+		BindingRuleExpectedNS string
+	}{
```

Comment:
Additional test cases should check that changing the mirroring prefix gets updated. That includes empty string to non-empty, non-empty to empty and non-empty to non-empty with different values.

Reply:
It would also be good to have one that tests a single destination namespace that is the Consul `default` namespace.

Reply:
> Additional test cases should check that changing the mirroring prefix gets updated. That includes empty string to non-empty, non-empty to empty and non-empty to non-empty with different values.

Added
* mirroring ns (no prefix) => mirroring ns (no prefix)
* mirroring ns => mirroring ns (same prefix)
* mirroring ns (no prefix) => mirroring ns (prefix)
* mirroring ns (prefix) => mirroring ns (no prefix)

> It would also be good to have one that tests a single destination namespace that is the Consul default namespace.

I added cases to `TestRun_ConnectInject_SingleDestinationNamespace`: `consulDestNamespaces := []string{"default", "destination"}`. How's that?

Reply:
Looks good!